service-base/src/main/java/org/eclipse/hono/service/registration/BaseDeviceRegistryServer.java

```diff
@@ -0,0 +1,134 @@
+     */
+    @Autowired
+    @Qualifier(AuthenticationConstants.QUALIFIER_AUTHENTICATION)
+    public void setAuthenticationServiceConnectionFactory(final ConnectionFactory factory) {
+        authenticationService = Objects.requireNonNull(factory);
+    }
+
+    @Override
+    protected void onRemoteConnectionOpen(final ProtonConnection connection) {
+        connection.setContainer(String.format("Hono-DeviceRegistry-%s:%d", getBindAddress(), getPort()));
+        setRemoteConnectionOpenHandler(connection);
+    }
+
+    @Override
+    protected void onRemoteConnectionOpenInsecurePort(final ProtonConnection connection) {
+        connection.setContainer(String.format("Hono-DeviceRegistry-%s:%d", getInsecurePortBindAddress(), getInsecurePort()));
+        setRemoteConnectionOpenHandler(connection);
+    }
+
+    private void setRemoteConnectionOpenHandler(final ProtonConnection connection) {
+        connection.sessionOpenHandler(remoteOpenSession -> handleSessionOpen(connection, remoteOpenSession));
+        connection.receiverOpenHandler(remoteOpenReceiver -> handleReceiverOpen(connection, remoteOpenReceiver));
+        connection.senderOpenHandler(remoteOpenSender -> handleSenderOpen(connection, remoteOpenSender));
+        connection.disconnectHandler(this::handleRemoteDisconnect);
+        connection.closeHandler(remoteClose -> handleRemoteConnectionClose(connection, remoteClose));
+        connection.openHandler(remoteOpen -> {
+            LOG.info("client [container: {}, user: {}] connected", connection.getRemoteContainer(), Constants.getClientPrincipal(connection).getName());
+            connection.open();
+            // attach an ID so that we can later inform downstream components when connection is closed
+            connection.attachments().set(Constants.KEY_CONNECTION_ID, String.class, UUID.randomUUID().toString());
+        });
+    }
+
+    private void handleSessionOpen(final ProtonConnection con, final ProtonSession session) {
+        LOG.info("opening new session with client [{}]", con.getRemoteContainer());
+        session.closeHandler(sessionResult -> {
+            if (sessionResult.succeeded()) {
+                sessionResult.result().close();
+            }
+        }).open();
+    }
+
+    /**
+     * Invoked when a client closes the connection with this server.
+     *
+     * @param con The connection to close.
+     * @param res The client's close frame.
+     */
+    private void handleRemoteConnectionClose(final ProtonConnection con, final AsyncResult<ProtonConnection> res) {
+        if (res.succeeded()) {
+            LOG.info("client [{}] closed connection", con.getRemoteContainer());
+        } else {
+            LOG.info("client [{}] closed connection with error", con.getRemoteContainer(), res.cause());
+        }
+        con.close();
+        con.disconnect();
+    }
+
+    private void handleRemoteDisconnect(final ProtonConnection connection) {
```

Comment:
I guess we can move this to `AmqpServiceBase` and make it *protected* so that subclasses can override it.